components/script/dom/bindings/record.rs

```diff
@@ -0,0 +1,211 @@
+    }
+
+    unsafe fn from_id(cx: *mut JSContext, id: HandleId) -> Option<DOMString> {
+        jsid_to_string(cx, id)
+    }
+}
+
+impl RecordKey for USVString {
+    fn to_utf16_vec(&self) -> Vec<u16> {
+        self.0.encode_utf16().collect::<Vec<_>>()
+    }
+
+    unsafe fn from_id(cx: *mut JSContext, id: HandleId) -> Option<USVString> {
+        rooted!(in(cx) let mut jsid_value = UndefinedValue());
+        let raw_id: RawHandleId = id.into();
+        JS_IdToValue(cx, *raw_id.ptr, jsid_value.handle_mut());
+
+        match USVString::from_jsval(cx, jsid_value.handle(), ()) {
+            Ok(ConversionResult::Success(s)) => Some(s),
+            Ok(ConversionResult::Failure(_)) => None,
+            Err(..) => None,
+        }
+    }
+}
+
+impl RecordKey for ByteString {
+    fn to_utf16_vec(&self) -> Vec<u16> {
+        self.iter().map(|&x| x as u16).collect::<Vec<u16>>()
+    }
+
+    unsafe fn from_id(cx: *mut JSContext, id: HandleId) -> Option<ByteString> {
+        rooted!(in(cx) let mut jsid_value = UndefinedValue());
+        let raw_id: RawHandleId = id.into();
+        JS_IdToValue(cx, *raw_id.ptr, jsid_value.handle_mut());
+
+        match ByteString::from_jsval(cx, jsid_value.handle(), ()) {
+            Ok(ConversionResult::Success(s)) => Some(s),
+            Ok(ConversionResult::Failure(_)) => None,
+            Err(..) => None,
+        }
+    }
+}
+
+/// The `Record` (open-ended dictionary) type.
+#[derive(Clone, JSTraceable)]
+pub struct Record<K: RecordKey, V> {
+    map: IndexMap<K, V>,
+}
+
+impl<K: RecordKey, V> Record<K, V> {
+    /// Create an empty `Record`.
+    pub fn new() -> Self {
+        Record {
+            map: IndexMap::new(),
+        }
+    }
+}
+
+impl<K: RecordKey, V> Deref for Record<K, V> {
+    type Target = IndexMap<K, V>;
+
+    fn deref(&self) -> &IndexMap<K, V> {
+        &self.map
+    }
+}
+
+impl<K, V, C> FromJSValConvertible for Record<K, V>
+where
+    K: RecordKey,
+    V: FromJSValConvertible<Config = C>,
+    C: Clone,
+{
+    type Config = C;
+    unsafe fn from_jsval(
+        cx: *mut JSContext,
+        value: HandleValue,
+        config: C,
+    ) -> Result<ConversionResult<Self>, ()> {
+        if !value.is_object() {
+            return Ok(ConversionResult::Failure(
+                "Record value was not an object".into(),
+            ));
+        }
+
+        rooted!(in(cx) let object = value.to_object());
+        let ids = IdVector::new(cx);
+        if !GetPropertyKeys(
+            cx,
+            object.handle(),
+            JSITER_OWNONLY | JSITER_HIDDEN | JSITER_SYMBOLS,
+            ids.get(),
+        ) {
+            // TODO: can GetPropertyKeys fail?
+            // (it does so if the object has duplicate keys)
+            // https://github.com/servo/servo/issues/21462
+            report_pending_exception(cx, false);
+            return Ok(ConversionResult::Failure(
+                "Getting Record value property keys failed".into(),
+            ));
```

Comment:
I don't understand this bit, why are we inhibiting the pending exception? Shouldn't we return `Err(_)`?

Reply:
Git blame says it's done in https://github.com/servo/servo/commit/74c1e00d8163f255bb4141ff3549bbdedd7ea766#diff-ff5f9b6a836f499ed5f16e31cc221a6aR69, @asajeffrey ?

Reply:
Seems @jdm [previously agreed to do this](https://github.com/servo/servo/pull/21029#issuecomment-408447065). Do you have some idea why do we have the inhibition here? Should we do the same for line 152?

Reply:
And maybe this should be updated too:

https://github.com/servo/servo/blob/0780fb0693d46ce321889e5222c5025b800415c2/components/script/dom/bindings/codegen/CodegenRust.py#L5272-L5273

Reply:
Fixed in b945205